### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by xiokd**
Code Pair: @dhoxie

### CODE REVIEW

**Comment by dhoxie** [comment]
## Review
- Create a *class library* project called "GenericsHomework.". ✔
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
  - Define the `Node` class
  - Include a constuctor that takes a value.  (No validation is necessary on the value). ✔
  - Add a `ToString()` override that writes out the value's `ToString()` result. ✔
  - Add a `Next` property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
    - The `Next` property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
    - The `Next` property setter should be private. ✔
  - Add an `Append` method that takes a value and appends a new `Node` instance after the current node (by invoking the `Next` property). ✔
  - Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
    - Whether it is sufficient to only set Next to itself ✔
    - Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
    - Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
  - Create an Exists method to test to see if a value exists in the list. ✔
  - Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔

## Extra Credit
We have not attempted extra credit

## Fundamentals
- Ensure you enable:
  - nullable reference types is enabled  ✔
  - net6 targetted  ✔
  - C# 10.0  ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

File: GenericsHomework/GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            while(currentNode.Next != this) { currentNode = currentNode.Next; }
+
+            currentNode.Next = newNode;
+            newNode.Next = this;
+        }
+
+        public void Clear()
+        {
+            // don't need to garbage collect as c# does so automatically
+            Next = this;
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework/Node.cs:36]: But how does C# know what to garbage collect automatically? Your implementation is correct, but I think you need to elaborate on it.

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+            Assert.AreEqual<string?>(null, result);
+        }
+
+        [TestMethod]
+        public void Append_GivenNewValues_Success()
+        {
+            Node<string> node = GetNodes();
+
+            Assert.IsTrue(node.Exists("Baggy"));
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:34]: Try writing a test that inserts multiple values

Comment (dhoxie) [reply GenericsHomework/GenericsHomework.Tests/NodeTests.cs:34]: The GetNodes() method inserts multiple values

**Pull request closed without merging**